src/sentry/workflow_engine/processors/data_source.py

```diff
@@ -0,0 +1,36 @@
+import sentry_sdk
+from django.db.models import Prefetch
+
+from sentry.utils import metrics
+from sentry.workflow_engine.models import DataPacket, DataSource, Detector
+
+
+def process_data_sources(
+    data_packets: list[DataPacket], query_type: DataSource.Type = DataSource.Type.SNUBA_QUERY
+) -> list[tuple[DataPacket, list[Detector]]]:
+    metrics.incr("sentry.workflow_engine.process_data_sources", tags={"query_type": query_type})
+
+    data_packet_ids = [packet.query_id for packet in data_packets]
+
+    # Fetch all data sources and associated detectors for the given data packets
+    with sentry_sdk.start_span(op="sentry.workflow_engine.process_data_sources.fetch_data_sources"):
+        data_sources = DataSource.objects.filter(
+            query_id__in=data_packet_ids, type=query_type
+        ).prefetch_related(Prefetch("datasourcedetector_set__detector", to_attr="detectors"))
```

Comment:
Nit: We could just not set `to_attr` i think, and just access this via `datasourcedetector_set`. Either way seems fine though